File: DirectProgramming/Fortran/Jupyter/OpenMP-offload-training/parallelism/main_test.f90
!==============================================================
! Copyright © 2020 Intel Corporation
!
! SPDX-License-Identifier: MIT
! =============================================================
program main
        use omp_lib
        integer, parameter :: ARRAY_SIZE=256, NUM_BLOCKS=9
        integer :: i,ib,correct_count=0, num_teams
        logical :: is_cpu = .true.
        real(kind=4) :: x(ARRAY_SIZE), y(ARRAY_SIZE) 
        real(kind=4) :: a = 1.5, tolerance=0.01
        double precision :: tb, te

        !initialize data
        do i=1,ARRAY_SIZE
                x(i) = i;
                y(i) = i;
        end do

        tb = omp_get_wtime() 

        !Perform Saxpy Function
        INCLUDE 'saxpy_func_parallel_solution.f90'

        te = omp_get_wtime() 
        
        print *, "Number of OpenMP Device Available:", omp_get_num_devices()
        if (is_cpu) then
                print *, "Running on CPU"
        else
                print *, "Running on GPU"
        end if
        print '("Work took ",(f7.5)," seconds.")', te-tb
        print *, "Number of Teams Created: ", num_teams
        !Check Results
        do i=1, ARRAY_SIZE
                if ( abs(y(i)-(a*i+i))<tolerance) correct_count=correct_count+1
        end do
        if (correct_count == ARRAY_SIZE) then
                print *, "Test: Passed"
        else
                print *, "Test: Failed"
        end if
end program main

File: DirectProgramming/C++/Jupyter/OpenMP-offload-training/parallelism/main_test.cpp
//==============================================================
// Copyright © 2020 Intel Corporation
//
// SPDX-License-Identifier: MIT
// =============================================================

#include <omp.h>
#include <stdio.h>

constexpr int ARRAY_SIZE = 256;
constexpr int NUM_BLOCKS = 9;

int main(int argc, char *argv[]) {
  int i, ib, is_cpu = 1, num_teams = 0;
  double tstart, tstop;
  float x[ARRAY_SIZE], y[ARRAY_SIZE];

  float a = 1.0f;
  float tolerance = 0.01f;
  int correct_count = 0;

  // Initialize some data
  for (i = 0; i < ARRAY_SIZE; i++) {
    x[i] = (float)i;
    y[i] = (float)i;
  }

  tstart = omp_get_wtime();

#include "saxpy_func_parallel_solution.cpp"

  tstop = omp_get_wtime();

  printf("Number of OpenMP Devices Available: %d\n", omp_get_num_devices());
  printf("Running on %s.\n", is_cpu ? "CPU" : "GPU");
  printf("Work took %f seconds.\n", tstop - tstart);
  printf("Number of Teams Created: %d\n", num_teams);
  for (int i = 0; i < ARRAY_SIZE; i++)
    if (abs(y[i] - (a * i + i)) < tolerance)
      correct_count++;
    else {
      printf("Incorrect Result at Element [%d] : %f\n", i, y[i]);
      break;
    }
  printf("Test: %s\n", (correct_count == ARRAY_SIZE) ? "PASSED!" : "Failed");
}
